File: src/features/history/fuzzy.ts
/**
 * Typo tolerance that does not turn into "everything matches".
 *
 * Two gates, in this order, because either alone is useless: trigram overlap
 * against the vocabulary generates candidates cheaply (no scan of the whole
 * term list per query term), then a *bounded* edit distance rejects the ones
 * that merely share letters. And it only fires when the term has almost no
 * exact postings.
 *
 * Expanded terms carry a penalty weight — but a weight is not a guarantee, and
 * this file cannot make one: BM25 saturation lets a strong hit on a commoner
 * neighbour out-earn a weak hit on the rare term the user actually typed. The
 * guarantee lives in rank.ts, which sorts every card containing a query term
 * above every card that only matched an expansion.
 */
import { type Bm25Index, documentFrequency, type WeightedTerm } from "./bm25.js";
import { editDistance } from "./edit-distance.js";
import { trigrams } from "./tokenize.js";

export { editDistance } from "./edit-distance.js";

export interface FuzzyOptions {
	/** Expand only a term with at most this many exact postings. */
	maxExactPostings?: number;
	/** Dice coefficient over trigrams a candidate must clear. */
	minOverlap?: number;
	/** How many expansions a single term may contribute. */
	maxCandidates?: number;
	/** Weight of a distance-1 expansion; distance-2 gets half of it. */
	penalty?: number;
}

/**
 * `minOverlap: 0.4` is what a transposition costs a short word — "retyr"
 * shares only 0.400 of its trigrams with "retry", so at 0.45 the commonest
 * typo class could never be corrected however cheap its edit distance was.
 *
 * Expanding a term with one posting is deliberate recall (the near-miss card
 * is still worth showing), and it is safe because the guarantee lives one
 * level up: rank.ts sorts every card that contains a query term ABOVE every
 * card that only matched an expansion, whatever the scores say.
 */
const DEFAULTS: Required<FuzzyOptions> = {
	maxExactPostings: 1,
	minOverlap: 0.4,
	maxCandidates: 3,
	penalty: 0.35,
};

/**
 * A gram shared by this much of the vocabulary discriminates nothing; walking
 * it would cost more than the whole query. Skipping it can only lose
 * candidates that a rarer gram of the same term also finds.
 */
const MAX_BUCKET = 4000;

/** Trigram → vocabulary terms, built once per index (indexes are immutable). */
const gramCache = new WeakMap<Bm25Index, Map<string, string[]>>();

function gramIndex(index: Bm25Index): Map<string, string[]> {
	const cached = gramCache.get(index);
	if (cached !== undefined) return cached;
	const map = new Map<string, string[]>();
	for (const term of index.postings.keys()) {
		for (const gram of new Set(trigrams(term))) {
			const bucket = map.get(gram);
			if (bucket === undefined) map.set(gram, [term]);
			else bucket.push(term);
		}
	}
	gramCache.set(index, map);
	return map;
}

/** Expansions for one term, best first; empty when the term needs no help. */
export function expandTerm(
	index: Bm25Index,
	term: string,
	options: FuzzyOptions = {},
): WeightedTerm[] {
	const opts = { ...DEFAULTS, ...options };
	if (documentFrequency(index, term) > opts.maxExactPostings) return [];
	const max = term.length <= 5 ? 1 : 2;
	const grams = new Set(trigrams(term));
	const shared = new Map<string, number>();
	const buckets = gramIndex(index);
	for (const gram of grams) {
		const bucket = buckets.get(gram);
		if (bucket === undefined || bucket.length > MAX_BUCKET) continue;
		for (const other of bucket) shared.set(other, (shared.get(other) ?? 0) + 1);
	}
	const scored: { term: string; distance: number; overlap: number }[] = [];
	for (const [candidate, hits] of shared) {
		if (candidate === term) continue;
		const overlap = (2 * hits) / (grams.size + trigrams(candidate).length);
		if (overlap < opts.minOverlap) continue;
		const distance = editDistance(term, candidate, max);
		if (distance > max) continue;
		scored.push({ term: candidate, distance, overlap });
	}
	scored.sort((a, b) => a.distance - b.distance || b.overlap - a.overlap);
	return scored
		.slice(0, opts.maxCandidates)
		.map((c) => ({ term: c.term, weight: opts.penalty / c.distance, fuzzy: true }));
}

/**
 * Exact terms at full weight, then penalised expansions for the weak ones.
 * Exact-first ordering is load-bearing: a term another query word expands to
 * must keep its full weight rather than be capped at the penalty.
 */
export function expandQuery(
	index: Bm25Index,
	terms: string[],
	options: FuzzyOptions = {},
): WeightedTerm[] {
	const exact = new Set(terms);
	const out: WeightedTerm[] = terms.map((term) => ({ term, weight: 1 }));
	for (const term of terms) {
		for (const expansion of expandTerm(index, term, options)) {
			if (!exact.has(expansion.term)) out.push(expansion);
		}
	}
	return out;
}

File: crates/flusk-core/src/index/fuzzy.rs
//! Port of src/features/history/fuzzy.ts: trigram candidates, then bounded
//! edit distance, only for terms with almost no exact postings. The ordering
//! guarantee (exact-containing cards above expansion-only cards) lives in
//! rank.rs, exactly as it does in the reference.

use super::bm25::{document_frequency, Bm25Index, WeightedTerm};
use super::edit_distance::edit_distance;
use super::tokenize::trigrams;
use std::collections::{HashMap, HashSet};

const MAX_EXACT_POSTINGS: usize = 1;
const MIN_OVERLAP: f64 = 0.4;
const MAX_CANDIDATES: usize = 3;
const PENALTY: f64 = 0.35;

/// A gram shared by this much of the vocabulary discriminates nothing.
const MAX_BUCKET: usize = 4000;

/// Trigram → vocabulary terms. The TS reference caches this per index in a
/// WeakMap; here the engine owns its index, so the map is built beside it.
pub struct GramIndex {
    buckets: HashMap<String, Vec<String>>,
}

impl GramIndex {
    pub fn build(index: &Bm25Index) -> Self {
        let mut buckets: HashMap<String, Vec<String>> = HashMap::new();
        // Deterministic vocabulary order: sorted, unlike JS Map insertion
        // order — harmless because candidates are re-sorted by (distance,
        // overlap) and ties beyond that cannot arise from bucket order alone.
        let mut terms: Vec<&String> = index.postings.keys().collect();
        terms.sort();
        for term in terms {
            let grams: HashSet<String> = trigrams(term).into_iter().collect();
            for gram in grams {
                buckets.entry(gram).or_default().push(term.clone());
            }
        }
        Self { buckets }
    }
}

/// Expansions for one term, best first; empty when the term needs no help.
pub fn expand_term(index: &Bm25Index, grams_of: &GramIndex, term: &str) -> Vec<WeightedTerm> {
    if document_frequency(index, term) > MAX_EXACT_POSTINGS {
        return Vec::new();
    }
    let max = if term.len() <= 5 { 1 } else { 2 };
    let grams: HashSet<String> = trigrams(term).into_iter().collect();
    let mut shared: HashMap<&str, usize> = HashMap::new();
    let mut order: Vec<&str> = Vec::new();
    for gram in &grams {
        let Some(bucket) = grams_of.buckets.get(gram) else { continue };
        if bucket.len() > MAX_BUCKET {
            continue;
        }
        for other in bucket {
            match shared.get_mut(other.as_str()) {
                Some(n) => *n += 1,
                None => {
                    shared.insert(other, 1);
                    order.push(other);
                }
            }
        }
    }
    let mut scored: Vec<(String, usize, f64)> = Vec::new();
    for candidate in order {
        if candidate == term {
            continue;
        }
        let hits = shared[candidate];
        let overlap = (2 * hits) as f64 / (grams.len() + trigrams(candidate).len()) as f64;
        if overlap < MIN_OVERLAP {
            continue;
        }
        let distance = edit_distance(term, candidate, max);
        if distance > max {
            continue;
        }
        scored.push((candidate.to_string(), distance, overlap));
    }
    scored.sort_by(|a, b| a.1.cmp(&b.1).then(b.2.partial_cmp(&a.2).unwrap_or(std::cmp::Ordering::Equal)));
    scored
        .into_iter()
        .take(MAX_CANDIDATES)
        .map(|(t, d, _)| WeightedTerm { term: t, weight: PENALTY / d as f64, fuzzy: true })
        .collect()
}

/// Exact terms at full weight first, then penalised expansions — exact-first
/// ordering is load-bearing (a term another word expands to keeps full weight
/// via score_terms' first-seen de-duplication).
pub fn expand_query(index: &Bm25Index, grams_of: &GramIndex, terms: &[String]) -> Vec<WeightedTerm> {
    let exact: HashSet<&String> = terms.iter().collect();
    let mut out: Vec<WeightedTerm> = terms
        .iter()
        .map(|t| WeightedTerm { term: t.clone(), weight: 1.0, fuzzy: false })
        .collect();
    for term in terms {
        for e in expand_term(index, grams_of, term) {
            if !exact.contains(&e.term) {
                out.push(e);
            }
        }
    }
    out
}
